### PR TITLE
Update Spark to 3.3.1

### PR DESCRIPTION
**Pull request by wangyum**
## Description

Spark 3.3.1 released. This PR upgrade spark to Spark 3.3.1.

## How was this patch tested?

Existing test.

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -35,6 +35,7 @@ lazy val commonSettings = Seq(
   fork := true,
   scalacOptions ++= Seq("-target:jvm-1.8", "-Ywarn-unused:imports"),
   javacOptions ++= Seq("-source", "1.8"),
+  resolvers += "Spark 3.3.1-rc1" at "https://repository.apache.org/content/repositories/orgapachespark-1418/",
```
Comment (wangyum) [inline build.sbt:38]: Will remove it if Spark 3.3.1 is officially released.

File: core/src/test/scala/org/apache/spark/sql/delta/MergeIntoSuiteBase.scala [hunk 2]
```diff
@@ -932,7 +932,8 @@ abstract class MergeIntoSuiteBase
 
       append(Seq((100, 100), (3, 5)).toDF("key2", "value"))
       // cache is in effect, as the above change is not reflected
-      checkAnswer(spark.table(s"delta.`$tempPath`"), Row(2, 2) :: Row(1, 4) :: Nil)
+      checkAnswer(spark.table(s"delta.`$tempPath`"),
+        Row(2, 2) :: Row(1, 4) :: Row(100, 100) :: Row(3, 5) :: Nil)
```
Comment (wangyum) [inline core/src/test/scala/org/apache/spark/sql/delta/MergeIntoSuiteBase.scala:936]: https://github.com/apache/spark/pull/37380 fixed this issue.

Comment (zsxwing) [reply core/src/test/scala/org/apache/spark/sql/delta/MergeIntoSuiteBase.scala:936]: Great to see this got fixed!

**Review by zsxwing — COMMENTED** [review]
Thanks for submitting the PR. Will merge this once 3.3.1 is out.

File: core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala [hunk 3]
```diff
@@ -127,9 +127,9 @@ class StatsCollectionSuite
         val statsDf = statsDF(deltaLog)
         assert(statsDf.where('numRecords.isNotNull).count() > 0)
         checkAnswer(statsDf, Seq(
-          Row(4, Row(10, 20), Row(17, 27)),
-          Row(2, Row(11, 21), Row(19, 29)),
-          Row(4, Row(12, 22), Row(18, 28))))
+          Row(3, Row(10, 20), Row(19, 29)),
+          Row(4, Row(12, 22), Row(17, 27)),
+          Row(3, Row(11, 21), Row(18, 28))))
```
Comment (wangyum) [inline core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala:132]: @zsxwing Is it a regression for Delta?

Comment (zsxwing) [reply core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala:132]: @wangyum do you know if there is any change in Spark `repartition`, such as changing the hash partitioning logic? Looks like the `repartition` behavior is slightly difference. Previously, we got 3 files and the numbers of records are  (4, 2, 4). But now we got a different distribution (3, 4, 3).

Comment (zsxwing) [reply core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala:132]: cc @cloud-fan

Comment (zsxwing) [reply core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala:132]: IIUC, the change here is a row may get a different hash value and write to a different file in 3.3.1?

Comment (wangyum) [reply core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala:132]: Looks like it's related to this commit: https://github.com/apache/spark/commit/2bae604d05ef734ed4562be5fc545d5e4abfbbd2

Comment (zsxwing) [reply core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala:132]: hm, this seems okay. I cannot think of any correctness issues for this change. Thanks for finding the commit.

Comment (zsxwing) [reply core/src/test/scala/org/apache/spark/sql/delta/stats/StatsCollectionSuite.scala:132]: I forgot that `repartition` is using `RoundRobinPartitioning`.

**Comment by wangyum** [comment]
cc @zsxwing

**Review by zsxwing — APPROVED** [review]
LGTM. Thanks!

**Pull request closed without merging**